[globalcache] Implement bi-directional support for serial devices

mhilbush opened the pull request:
Implements bi-directional support for GlobalCache devices with serial ports (GC-100, iTach Flex, and iTach SL).
Fixes issues from code analysis

Signed-off-by: Mark Hilbush <mark@hilbush.com>

martinvw commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/GlobalCacheBindingConstants.java:
```diff
@@ -25,7 +25,7 @@
     public static final String BINDING_ID = "globalcache";
 
     // Maximum number of GC devices supported by the binding
```
The comment is no longer in line with the code?

martinvw commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/GlobalCacheBindingConstants.java:
```diff
@@ -63,6 +63,12 @@
     public static final String ACTIVE_CABLE_RELAY = "FLEX_RELAY";
 
+    // Serial readers
+    public static final String ENABLE_TWO_WAY_1 = "enableTwoWay1";
```
Maybe use a few extra characters everywhere like `ENABLE_TWO_WAY_PORT_1` etc, it sounds more sensible than `ENABLE_TWO_WAY_1`. So add the fact that you are referring to a property for either port 1 or port 2.

martinvw commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java:
```diff
@@ -284,8 +292,8 @@ private boolean isHexCode(String code) {
      */
     private String convertHexToGC(String hexCode) throws HexCodeConversionException {
         // Magic number for converting frequency to GC format
-        final int FREQ_CONVERSION_FACTOR = 4145146;
```
You could also make them real constants, now you might waste 0,000001% performance by initializing them every call.

mhilbush replied:
The cost of initialization seems insignificant when compared to the rest of the method.

martinvw replied:
Agree on that, see my very accurate gues of the performance waste 😉

martinvw commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java:
```diff
@@ -506,49 +529,46 @@ public void run() {
+        private void writeSerialToDevice(RequestMessage requestMessage) throws IOException {
+            DataOutputStream out = connectionManager.getSerialOut(requestMessage.getCommandType());
+            if (out == null) {
+                logger.error("Can't send serial command; output stream is null!");
```
Make this a warn
Should the thing status be updated?

martinvw commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java:
```diff
@@ -565,6 +585,9 @@ private String readReplyFromDevice(RequestMessage requestMessage) throws IOExcep
         private DeviceConnection serial1Connection;
         private DeviceConnection serial2Connection;
 
+        private SerialReader serialReader1;
```
I assume these one and two also refer to the ports, please change the naming here also like that.

martinvw commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java:
```diff
@@ -621,89 +644,119 @@ protected void connect() {
         }
 
-        private boolean deviceConnect(DeviceConnection conn) {
+        private boolean commandConnect(DeviceConnection conn) {
```
Why this name change?

mhilbush replied:
I wanted to clearly differentiate between the connection for handling globalcache commands versus the connection(s) for the serial port(s).  The command port and serial port(s) have distinctly different functionality, so I thought it best to use different terminology.

martinvw replied:
Thanks for the clarification

martinvw commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java:
```diff
@@ -621,89 +644,119 @@ protected void connect() {
                         thingID(), conn.getIP(), e.getMessage());
                 markThingOfflineWithError(ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, e.getMessage());
-                deviceDisconnect(conn);
+                commandDisconnect(conn);
```
Are you calling this method again recursively on an IOException, this might explode into a StackoverFlowError or do I missing something?

mhilbush replied:
I don't think so.  It should just result in closing the socket that was opened above.  I'll check it again.

martinvw replied:
Were you able to check this?

mhilbush replied:
Looked at it again this AM.  It looks ok to me.

martinvw replied:
So we if get an IOException over and over again we will not recursively call it until it explodes? Why not what stoppes it from doing so?

mhilbush replied:
Maybe I'm missing something (and I would not rule that out), but I'm not seeing the recursion.  It would be a bit unusual to not get the streams immediately after successfully opening the socket.  Nonetheless, when this IOException happens, the socket will be closed (deviceDisconnect) and the connection will be marked as disconnected.  An IOException could occur when the socket is closed, and that will be caught and logged.  On the next connection attempt, if the condition that caused the IOException persists, the above scenario would occur again.  But, I don't see where the recursion occurs.

Having said that, I do see a separate issue that I'm going to look at that I think has been there since the initial release.

martinvw replied:
Its just me being blind :-S it was indeed disconnecting from the connect mode, nothing recursive going on.

Sorry!

mhilbush replied:
No worries.  Making me look at the code more closely had a good outcome, as there is something else I need to fix.

martinvw commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java:
```diff
@@ -788,10 +856,201 @@ private void checkConnection() {
+            Boolean enableTwoWay = (Boolean) thing.getConfiguration().get(enableTwoWayConfig);
+            logger.debug("Enable two-way is {} for thing {} {}", enableTwoWay, thingID(), serialDevice);
+
+            if (enableTwoWay != null && enableTwoWay.equals(Boolean.TRUE)) {
```
Position literals first in String comparisons - that way if the Object is null you won't get a NullPointerException and don't need an extra check, it'll just return false. [source: PMD](http://pmd.sourceforge.net/pmd-4.3.0/rules/design.html#PositionLiteralsFirstInComparisons)

martinvw commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java:
```diff
@@ -788,10 +856,201 @@ private void checkConnection() {
+
+            if (enableTwoWay != null && enableTwoWay.equals(Boolean.TRUE)) {
+                // Get the end of message delimiter from the config, URL decode it, and convert it to a byte array
+                String eomString = (String) thing.getConfiguration().get(eomDelimiterConfig);
```
Can you replace eom by endOfMessage I know its long, but I had no idea what eom means and do not know what the long version means.

martinvw commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java:
```diff
@@ -788,10 +856,201 @@ private void checkConnection() {
+        }
+
+        private void stopSerialReaders() {
+            if (deviceSupportsSerial1() && serialReader1 != null) {
```
1) Is the second half of the if enough?
2) should you `null` it out?

mhilbush replied:
1. I wanted to be consistent with how I was doing it in the method to start the readers.
2. Yes

martinvw commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java:
```diff
@@ -788,10 +856,201 @@ private void checkConnection() {
+        }
+
+        public void stop() {
+            if (serialReaderJob != null) {
```
Should the `BufferedInputStream` be closed?

mhilbush replied:
The streams are opened and closed in one place (serialConnect and serialDisconnect).

martinvw replied:
Clear!

martinvw commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java:
```diff
@@ -788,10 +856,201 @@ private void checkConnection() {
+
+                } catch (UnsupportedEncodingException e) {
+                    logger.info("Unsupported encoding exception: {}", e.getMessage(), e);
+                    continue;
```
2x continue is not needed as the last statement for this loop

mhilbush replied:
I debated this.  I came down on the side of being explicit, so as to not make one wonder if I overlooked that catch clause compared to the other 2 below it.  I can go either way on it, so I'll defer to your best judgement.

mhilbush commented:
@martinvw I made several changes based on your feedback.  Thanks.

I also commented on a few of your review comments.  Let me know your thoughts on those.

martinvw commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java:
```diff
@@ -788,10 +856,202 @@ private void checkConnection() {
+            return buf.toByteArray();
+        }
+
+        private boolean findEOM(byte[] buf, byte[] endOfMessage) {
```
Should we also rename this one

mhilbush replied:
Ah, missed that one.  :-(  I'm gonna leave lengthEOM alone because I think it's clear given the context.

martinvw replied:
Yes I agree on that!

martinvw commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java:
```diff
@@ -788,10 +856,202 @@ private void checkConnection() {
+        }
+
+        private void stopSerialReaders() {
+            if (deviceSupportsSerial1() && serialReaderPort1 != null) {
```
Should we also include the port here

mhilbush replied:
Yes

martinvw commented:
Thanks!

kaikreuzer commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java:
```diff
@@ -66,14 +71,17 @@
     private CommandProcessor commandProcessor;
-    ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(MAX_GC_DEVICES);
-    ScheduledFuture<?> scheduledFuture;
+    private ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(MAX_GC_THREADS);
```
I see that you didn't introduce it in this PR, but in general, I would advice to use `ThreadPoolManager.getScheduledPool` as this also nicely names the threads and allows people to reconfigure the pool size (default is 5, which should work here as well, right?).

mhilbush replied:
I wasn't aware I could use this to get my own dedicated pool.  Thanks!

I'll rework this.  And, yes, 5 should be enough (command processor, serial port 1 reader, serial port 2 reader, and connection monitor threads).  Ok to do the rework in this PR?

I'll need to do this in one of my other bindings, as well.

kaikreuzer replied:
> Ok to do the rework in this PR?

Yes, sure!

mhilbush replied:
One question on this.  Will I need to have a separately named pool for each handler instance, since each handler instance can have up to three permanent threads, as well as one thread that runs intermittently?  Or, is there another approach I should take?

kaikreuzer replied:
If you need that many threads for every single handler (sigh...), I guess it makes most sense to indeed create a separately named pool for each, because otherwise you will run out of available threads quickly if you have more than a single handler.

mhilbush replied:
I think using that many threads per handler will be more the exception rather than the rule.  I say that because the only device that will use 3 permanent threads is the GC-100-12 (which I suspect is the least prevalent model).  The iTach SL, GC-100-6, and Flex (when configured for serial) will use 2 permanent threads.  The other devices (iTach CC, iTach IR, Flex (configured for IR or CC), and the zMote) all will use 1 permanent thread.

I could be smarter at initialization time to create the thread pool with a size that is more appropriate for the device type; however, I don't see a way to do that with the current API.

kaikreuzer replied:
The thread pools reduce their size down to 1 if no threads are required - so having multiple pools is no issue, they won't spawn 5 threads each, if not needed.

mhilbush replied:
Cool.  Then I think one pool per handler will solve the problem and not be wasteful.

kaikreuzer commented on addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java:
```diff
@@ -264,7 +272,7 @@ private String lookupCode(Command command, ChannelUID channelUID) {
                 logger.debug("Converted hex code is: {}", code);
             } catch (HexCodeConversionException e) {
-                logger.warn("Failed to convert hex code to GC format");
+                logger.info("Failed to convert hex code to GC format");
```
That's a weird message for an info. I would either expect it to be some bug and thus would want to see an error/warning including stacktrace or to only have that as debug.

mhilbush replied:
I think it should stay as info, as it's a user "data" issue.  The user can put two different IR code formats in the MAP file - either a hex IR code or a globalcache IR code.  If it's a hex IR code, it needs to be converted to globalcache format using the method `convertHexToGC`.  That conversion could fail for a number of reasons, all of which will generate a `HexCodeConversionException`.  I do need to include the message from the exception, so that the user knows what aspect of the hex code caused the conversion failure so that they can fix it.  Ok?

kaikreuzer replied:
ok... This whole binding is anyhow pretty exceptional in terms of "low level functionality" - normally, a user should never be required to check the logs if the issue is within his configuration. :-)

mhilbush replied:
Yes, messing with IR codes can be... well, I guess a bit messy.  :-)

I suppose I could update the thing status to show it online, but with a detailed status message (I think that's possible?).  But that didn't seem any better solution than logging an info level message.

mhilbush commented:
I believe I've addressed all the comments so far, as well as fixed an issue that @martinvw helped me uncover.

Thanks for the feedback so far.  Let me know if there's anything else I need to do.

kaikreuzer commented:
Many thanks!
